### PR TITLE
pe-newstyle-factor-35

### PR DESCRIPTION
**Pull request by pauleveritt**
First draft of the factory proposal's science fiction.

### CODE REVIEW

File: docs/factories.rst [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+        return Greeting(customer=customer)
+
+Injector DSL: Attributes
+========================
+
+The earlier ``wired`` dataclass injector solved an additional problem.
+
+Our ``Greeting`` gets a ``Customer``.
+It can then have a method that says hello, e.g. ``f'Hello {self.customer.name}'``.
+But that's a big surface are -- the entire ``Customer`` -- when the contract really only wants one piece of data.
```
Comment (mmerickel) [inline docs/factories.rst:251]: area

File: docs/factories.rst [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+.. code-block:: python
+
+    @component()
+    @dataclass
+    class Greeting:
+        name: str
+        punctuation: Injected[str, Config, Attr('punctuation')]
+
+        def __call__(self) -> VDOM:
+            return html('<div>Hello {self.name}{self.punctuation}'
```
Comment (mmerickel) [inline docs/factories.rst:341]: missing `)`

File: docs/factories.rst [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+
+Let's use PEP 593 "Flexible Annotations".
+``wired`` would add:
+
+.. code-block:: python
+
+    Injector = object()
+    LT = TypeVar("local_type")
+    RT = TypeVar("registered_type")
+    Injected = Annotated[LT, RT, Injector]
```
Comment (mmerickel) [inline docs/factories.rst:207]: This is getting pretty wild for me as someone who is barely even using typing. It's interesting though!

**Comment by pauleveritt** [comment]
@mmerickel I split out the science fiction for factory and injector. We can focus on the former. Jim has offered to help me on some decorator and Signature stuff.

The last two sections in `factories.rst` have some changed points which I'm curious whether you are ok on. It's about functions instead of classes.

**Comment by pauleveritt** [comment]
I pushed an implementation of the science fiction with a bunch of tests.

File: docs/factories.rst [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+            ...
+
+With ``wired``, you have a factory function that can get information from the container, construct an instance, and return it:
+
+.. code-block:: python
+
+    def login_factory(container):
+        return LoginService()
+
+In the existing ``wired``, registering a factory means is a method call on the registry:
```
Comment (mmerickel) [inline docs/factories.rst:30]: "means is a" ?

File: docs/factories.rst [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+For such a simple case, the ``login_factory`` function isn't really needed.
+In fact, several pieces can be streamlined, especially in the presence of decorator support and type hinting.
+Let's take a look.
+
+The Simplest Case
+=================
+
+The above -- contrived -- example just constructs a class.
+Meaning, the class *constructor* is entirety of the factory.
```
Comment (mmerickel) [inline docs/factories.rst:51]: is "the" entirety

File: docs/factories.rst [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+Imagine eliminating the ``def login_factory`` function and just doing this directly:
+
+.. code-block:: python
+
+    registry.register_service(LoginService)
```
Comment (mmerickel) [inline docs/factories.rst:56]: This is taking things a step too far and coupling the interface/context/name into a "service descriptor" which is a combination of `(factory, interface, context, name)`. I thought the plan was just to focus on the factory part of that and making it easier to define.

**Review by mmerickel — CHANGES_REQUESTED** [review]

**Comment by pauleveritt** [comment]
As discussed in chat, I will retract this PR. Individual pieces might resurface later in easier-to-digest chunks.

**Pull request closed without merging**